### PR TITLE
Update Zenoh-pico from 0.7.0 to 1.0.0

### PR DESCRIPTION
**Pull request by AlexisTM**
### Solved Problem

Zenoh 1.0.0 is soon to be released. This upgrades Zenoh-pico version from 0.7.0 to 1.0.0.

The API of Zenoh won't be changing anymore, making the upgrades trivial from now on. This also means from this version onwards, we should feel comfortable to add features without having to re-implement the Zenoh interface from scratch.

Followup of https://github.com/PX4/PX4-Autopilot/pull/22017

### Solution
- Updates the zenoh-pico version. It still needs one update on Zenoh-pico side (https://github.com/eclipse-zenoh/zenoh-pico/pull/561) to be directly usable, which is mirrored with the fix in our [px4 fork](https://github.com/PX4/zenoh-pico/tree/dev/1.0.0-px4).
- Updates the usage of API of Zenoh-pico within the PX4 Zenoh module
- Enables /dev/urandom device driver **for px4fmu-v6x**

### Changelog Entry
For release notes:
```
Feature Zenoh 1.0
Updated the Zenoh version to 1.0.
Enables /dev/urandom for fmu-v6x
```

### Usage

On your machine, tested in ROS 2 Jazzy
```
# The dds plugin uses cyclonedds serialization
sudo apt install ros-jazzy-rmw-cyclonedds-cpp
# Netplan config of your ethernet https://docs.px4.io/main/en/advanced_config/ethernet_setup.html
# I used 10.41.10.1/24 on a USB Ethernet adapter

# Install the zenoh ros2dds plugin to forward the topics
mkdir -p ~/ros2/src
cd ~/ros2/src
git clone https://github.com/eclipse-zenoh/zenoh-plugin-ros2dds/ -b dev/1.0.0
git clone https://github.com/PX4/px4_msgs
cd ~/ros2
colcon build

# Start the Zenoh router with ros2dds
source ~/ros2/install/setup.bash
export RMW_IMPLEMENTATION=rmw_cyclonedds_cpp
RUST_LOG=debug zenoh_bridge_ros2dds -l tcp/10.41.10.1:7447
```

PX4 setup
```
make px4-fmuv6x_zenoh upload
# Netman config https://docs.px4.io/main/en/advanced_config/ethernet_setup.html

# in NSH
zenoh config net client tcp/10.41.10.1:7447
zenoh config addpublisher vehicle_status vehicle_status
zenoh config addpublisher sensor_combined sensor_combined
zenoh start
```

Receive the data
```
source ~/ros2/install/setup.bash
export RMW_IMPLEMENTATION=rmw_cyclonedds_cpp
ros2 topic echo /rt/sensor_combined px4_msgs/SensorCombined
ros2 topic echo /rt/vehicle_attitude px4_msgs/VehicleAttitude
```

### Context

This is still to be considered as highly experimental.

NOTE: Zenoh requires `/dev/urandom` to be available which means you should use `make [platform] menuconfig` and enable it under `device drivers` on the platform you want to use it on.

### Limitations

- PX4 crashes if the link goes down (at least in client mode)
- I failed to use the peer mode (multicast mode)
- The ROS2 DDS plugin requires the message type to be available to use `ros2 topic echo`
- No timesync
- No rate limiting
- Naive CSV file for configuration, having a common configuration file between microXRCE and Zenoh would be great
- FMU-to-FMU communication no distinction between own msgs and from incoming from Zenoh
- No documentation yet
- No message version checking - Different versions of PX4 may produce corrupted data (px4_msgs & px4 firmware sync)

### CODE REVIEW

File: .gitmodules [hunk 1]
```diff
@@ -70,8 +70,8 @@
 	url = https://github.com/px4/rosidl
 [submodule "src/modules/zenoh/zenoh-pico"]
 	path = src/modules/zenoh/zenoh-pico
-	url = https://github.com/px4/zenoh-pico
-	branch = pr-zubf-werror-fix
+	url = https://github.com/eclipse-zenoh/zenoh-pico
```
Comment (PetervdPerk-NXP) [inline .gitmodules:73]: I've pushed dev/1.0.0 to the PX4 fork.
could you change url back to px4/zenoh-pico

**Comment by PetervdPerk-NXP** [comment]
Great work!
Could you run `make format` to fix the style issues.
Also did you test the multicast option as well?
One nice thing about zenoh is that 2 PX4 instances can communicate with each-other.

File: .gitmodules (hunk 1 above)
Comment (AlexisTM) [reply .gitmodules]: This has been updated

**Comment by PetervdPerk-NXP** [comment]
Btw it seems that for some odd reason `zenoh-bridge-dds` isn't compatible by design with `rmw_zenoh`

https://github.com/ros2/rmw_zenoh/issues/201
> The intended use case of this rmw is not to communicate with other native Zenoh applications

Which is quite unfortunate but I think it's fixable, the whole is goal of using Zenoh on PX4 is that we can work without a bridge and we can do p2p multicast. communication.

Looking at this PR https://github.com/ros2/rmw_zenoh/pull/171 we need to change the keyexpr to be compatible with `rmw_zenoh`
`<ros_domain_id>/<topic_name>/<topic_type>/<topic_hash>`

**Comment by AlexisTM** [comment]
I looked further. The sensor_combined topic actually works now, I built it all from scratch and probably had some options clashing last week.

I didn't dig rmw_zenoh as it is not on the correct version and there are over 300 `z_` functions to review to go for 1.0.0. 

I tried using [ros2dds](https://github.com/eclipse-zenoh/zenoh-plugin-ros2dds) as well as the [dds](https://github.com/eclipse-zenoh/zenoh-plugin-dds) compatibility layer with DDS without success as they both report: 
```
(venv) ➜  PX4-Zenoh git:(pr-zenoh-1.0) ✗ ros2 topic echo /rc/vehicle_status px4_msgs/VehicleStatus
The passed message type is invalid
```

On the other hand, I successfully decode them manually using `zenoh-python` and `idlc`.

**Comment by PetervdPerk-NXP** [comment]
You could try if the official zenoh pico dds example also has this problem.
https://github.com/eclipse-zenoh/zenoh-demos/blob/main/zenoh-pico-dds/helloworld/z_pub_cdr.c
Then it would mean that something is wrong with the ros2dds zenoh plugin.

**Comment by AlexisTM** [comment]
Thank you for the hint! I will try that tomorrow :)

If you want to test the Python interface through idlc, I intentionally pushed the px4 idl file & generated library.

https://github.com/AlexisTM/zenoh-python/blob/df8029845409a3b304ba103dd4d67c00ee4819ba/examples/z_sub_px4.py#L100-L110

image

**Comment by AlexisTM** [comment]
@PetervdPerk-NXP Actually, the ROS 2 dds part works. It was my bad as I installed `ros-jazzy-desktop` which I supposed (wrongly) was including `rmw_cyclonedds` as it is the default rmw.

I will now update the description.

**Comment by PetervdPerk-NXP** [comment]
```ros2 topic echo /rt/sensor_combined px4_msgs/SensorCombined```

Do you have to add `rt/` to be able to subscribe on ROS2?
I would expect the following to work
```ros2 topic echo sensor_combined px4_msgs/SensorCombined```
ROS2 should take care of it.

**Comment by AlexisTM** [comment]
I just tried it and confirm I need to use the namespace `rt`.

**Comment by PetervdPerk-NXP** [comment]
> My guess is I have been using zenoh-bridge-ros2dds but before you were using zenoh-bridge-dds. As the rt/ is a quirk of ROS 2, ros2dds adds the prefix automagically. We are publishing (with ros2dds) to rt/rt/xxx

Indeed ros2dds didn't exist yet when I was implementing it.
I'm okay with removing the `rt/` on our side.

https://github.com/PX4/PX4-Autopilot/blob/dc7b77e2919e2dff86d3191ffd7e27a2abe735eb/src/modules/zenoh/publishers/zenoh_publisher.hpp#L72-L75

https://github.com/PX4/PX4-Autopilot/blob/dc7b77e2919e2dff86d3191ffd7e27a2abe735eb/src/modules/zenoh/publishers/zenoh_publisher.cpp#L64-L74

https://github.com/PX4/PX4-Autopilot/blob/dc7b77e2919e2dff86d3191ffd7e27a2abe735eb/src/modules/zenoh/subscribers/zenoh_subscriber.hpp#L76-L79

https://github.com/PX4/PX4-Autopilot/blob/dc7b77e2919e2dff86d3191ffd7e27a2abe735eb/src/modules/zenoh/subscribers/zenoh_subscriber.cpp#L81-L91

**Comment by AlexisTM** [comment]
I removed the `rt/` prefix (as well as the rostopic now unused switch) and we can now subscribe to the topic without the `rt/` prefix on ros side.

**Review by PetervdPerk-NXP — APPROVED** [review]

**Comment by PetervdPerk-NXP** [comment]
I'm okay with merging this as is, hopefully rmw_zenoh will move soon to 1.0.0 as well.
Could update the instruction in the PR description without the `rt/` prefix.

**Review by dagar — APPROVED** [review]

**Comment by MaEtUgR** [comment]
image
on main probably because
image
I guess the branch `dev/1.0.0-px4` (https://github.com/PX4/zenoh-pico/commits/dev/1.0.0-px4/) was updated but the submodule commit stayed on the previous commit not containing https://github.com/eclipse-zenoh/zenoh-pico/pull/560
image

**Comment by MaEtUgR** [comment]
Is this what you wanted? https://github.com/PX4/PX4-Autopilot/pull/23492

**Comment by bake1912** [comment]
Hello PX4 Team,
I am encountering an issue while running the following setup:
Command: make px4_sitl_zenoh gz_x500
PX4 Commit Hash: 058fe540dd05b600f2df07104ef031a57b229c42
I added the following Zenoh publishers:

```
zenoh config addpublisher fmu/out/vehicle_status vehicle_status
zenoh config addpublisher fmu/out/vehicle_global_position vehicle_global_position
zenoh config addpublisher fmu/out/vehicle_odometry vehicle_local_position
zenoh config addpublisher fmu/out/sensor_combined sensor_combined
```
Everything works fine when I keep the number of Zenoh publishers under 9. However, when I try to add more publishers, I encounter a segmentation fault like
``` 
Error CDR buffer is too small Segmentation fault.
```

Could you please guide me on how to debug or resolve this issue? Is there a known limitation with the number of Zenoh publishers, or could it be related to memory allocation?

Looking forward to your advice!
Thank you

**Pull request merged**